source/module_base/test/matrix_test.cpp

```diff
@@ -0,0 +1,334 @@
+#include"../matrix.h"
+#include"gtest/gtest.h"
+
+/************************************************
+*  unit test of class matrix and related functions
+***********************************************/
+
+/**
+ *  - Tested functions of class matrix:
+ *      - constructor:
+ *          - constructed by nrow and ncloumn
+ *          - constructed by a matrix
+ *          - constructed by the rvalue of a matrix
+ *      - function create
+ *      - operator "=": assigned by a matrix or the rvalue of a matrix
+ *      - operator "()": access the element
+ *      - operator "*=", "+=", "-="
+ *      - function trace_on
+ *      - function zero_out
+ *      - function max/min/absmax
+ *      - function norm
+ *      - function print (not called in abacus, no need to test)
+ * 
+ * - Tested functions related to class matrix
+ *      - operator "+", "-", "*" between two matrixs
+ *      - operator "*" between a double and a matrix, and reverse.
+ *      - function transpose
+ *      - function trace_on
+ *      - function mdot
+ * 
+ */
+
+//a mock function of WARNING_QUIT, to avoid the uncorrected call by matrix.cpp at line 37.
+namespace ModuleBase
+{
+    void WARNING_QUIT(const std::string &file,const std::string &description) {return ;}
+}
+
+class matrixTest : public testing::Test
+{
+    protected:
+    ModuleBase::matrix m23a,m33a,m33b,m33c,m34a,m34b;
+
+    void SetUp()
+    {
+        m23a.create(2,3);
+        for (int i=1;i<=6;++i) {m23a.c[i-1] = i*1.0;}
+
+        m33a.create(3,3);
+        for (int i=1;i<=9;++i) {m33a.c[i-1] = i*1.0;}
+
+        m33b.create(3,3);
+        for (int i=1;i<=9;++i) {m33b.c[i-1] = i*11.1;}
+
+        m33c.create(3,3,true);
+        m34a.create(3,4,true);
+        m34b.create(3,4,true);
+    }
+
+};
+
+TEST(matrix,Constructornrnc)
+{
+    ModuleBase::matrix m(3,4,true);
+    EXPECT_EQ(m.nr,3);
+    EXPECT_EQ(m.nc,4);
+    EXPECT_DOUBLE_EQ(m(0,0),0.0);
+    //EXPECT_DEATH(ModuleBase::matrix m(0,1),"");
```

Comment:
Why commented out? What is the expected behavior if calling with a non-positive integer??

Reply:
> Why commented out? What is the expected behavior if calling with a non-positive integer?

Reply:
Previous, I think the program should be terminated when construct a matrix by using a non-positive integer.  But, now, i think no one would use it as this case. So I will delete it.

Reply:
> 



> Previous, I think the program should be terminated when construct a matrix by using a non-positive integer. But, now, i think no one would use it as this case. So I will delete it.

Agree. In practice, checking array access out of bound everytime results large overhead.
However, there are some differece between it and C++ convention. C++ array starts with 0, and this feature ensures 
 if you pass an index of type `size_t`(unsigned), [0] will access the first element.